rust/worker/src/execution/orchestration/compact.rs

```diff
@@ -812,18 +818,69 @@ impl Handler<TaskResult<ApplyLogToSegmentWriterOutput, ApplyLogToSegmentWriterOp
         };
 
         if num_tasks_left == 0 {
-            let segment_writer = self.get_segment_writer_by_id(message.segment_id).await;
-            let segment_writer = match self.ok_or_terminate(segment_writer, ctx) {
-                Some(writer) => writer,
+            let writers = self.get_segment_writers().await;
+            let writers = match self.ok_or_terminate(writers, ctx) {
+                Some(writers) => writers,
                 None => return,
             };
 
-            self.dispatch_segment_writer_commit(segment_writer, ctx.receiver(), ctx)
-                .await;
+            if message.segment_id == writers.metadata.id {
+                let span = self.get_segment_writer_span(&ChromaSegmentWriter::MetadataSegment(
+                    writers.metadata.clone(),
+                ));
+                let operator = PrefetchForMetadataWriterOperator::new();
+                let input = PrefetchForMetadataWriterInput::new(
+                    self.blockfile_provider.clone(),
+                    writers.metadata,
+                );
+                let task = wrap(Box::new(operator), input, ctx.receiver());
+                let res = self.dispatcher().send(task, Some(span)).await;
+                self.ok_or_terminate(res, ctx);
+            } else {
+                let segment_writer = self.get_segment_writer_by_id(message.segment_id).await;
+                let segment_writer = match self.ok_or_terminate(segment_writer, ctx) {
+                    Some(writer) => writer,
+                    None => return,
+                };
+
+                self.dispatch_segment_writer_commit(segment_writer, ctx.receiver(), ctx)
+                    .await;
+            }
         }
     }
 }
 
+#[async_trait]
+impl Handler<TaskResult<PrefetchForMetadataWriterOutput, PrefetchForMetadataWriterError>>
+    for CompactOrchestrator
+{
+    type Result = ();
+
+    async fn handle(
+        &mut self,
+        message: TaskResult<PrefetchForMetadataWriterOutput, PrefetchForMetadataWriterError>,
+        ctx: &ComponentContext<CompactOrchestrator>,
+    ) {
+        match self.ok_or_terminate(message.into_inner(), ctx) {
+            Some(message) => message,
+            None => return,
+        };
+
+        let writers = self.get_segment_writers().await;
+        let writers = match self.ok_or_terminate(writers, ctx) {
+            Some(writers) => writers,
+            None => return,
+        };
+
+        self.dispatch_segment_writer_commit(
```

Comment:
oo, the pattern elsewhere is that prefetch is fire and forget. So `dispatch_segment_writer_commit` task is dispatched immediately after dispatching the prefetch task